tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -651,6 +615,52 @@ void promoteToSharedGreedy(
     scop.insertSyncsAroundCopies(bandNode);
   }
 }
+
+/*
+ * Check that accesses described by "group" are thread-private within the given
+ * scope, i.e. that no two threads access the same tensor element.  The scope
+ * is defined by "prefixSchedule" that describes the outer loops.  Within
+ * different instances of "prefixSchedule", threads may access the same
+ * element.  It is expected to be copied to registers at the beginning of the
+ * scope, and back to some location accessible to all threads at the end of the
+ * scope.  If synchronization is necessary due to data dependences, it should
+ * have been inserted by the mapping at the end of the scope.
+ */
+bool isThreadPrivate(
+    const TensorReferenceGroup& group,
+    isl::multi_union_pw_aff prefixSchedule,
+    isl::multi_union_pw_aff threadSchedule) {
+  auto accesses = group.originalAccesses(); // D -> A
+  auto prefixScheduleUmap = isl::union_map::from(prefixSchedule); // D -> Sp
+  auto accessesWithinScope =
+      accesses.range_product(prefixScheduleUmap); // D -> (Sp -> A)
+  auto threadScheduleUmap = isl::union_map::from(threadSchedule); // D -> St
+  auto threadAccesses =
+      accessesWithinScope.apply_domain(threadScheduleUmap); // T -> (Sp -> A)
+  return threadAccesses.is_injective();
+}
+
+/*
+ * Check if the tensor reference group "group" can be promoted to registers at
+ * the given "scope" in the schedule tree rooted at "root".  "prefixSchedule"
+ * is the schedule above the scoping point, "scope" inclusive.
+ * "threadSchedule" is the thread mapping schedule covering all points in the
+ * subtree rooted at "scope".
+ *
+ * In practice, check that no elements in this group are accessed by different
+ * threads and that all loop indices that may appear in subscripts correspond
+ * to unrolled loops.
+ */
+inline bool isPromotableToRegistersBelow(
+    const TensorReferenceGroup& group,
+    const detail::ScheduleTree* root,
+    const detail::ScheduleTree* scope,
+    isl::multi_union_pw_aff prefixSchedule,
+    isl::multi_union_pw_aff threadSchedule) {
+  return isThreadPrivate(group, prefixSchedule, threadSchedule) &&
+      accessSubscriptsAreUnrolledLoops(
+             group, root, scope, prefixSchedule, threadSchedule);
+}
```

Comment:
Did you just rename `isPromotableToRegisterBelowThreads` to `isPromotableToRegistersBelow` and then split it up into two functions?
Surely, that could have been done in a separate commit.

Reply:
No, I wrote a completely new `isPromotableToRegistersBelow` (including the unrolling part that I separated in a different commit) and then called it from `isPromotableToRegistersBelowThreads`.

Reply:
On Wed, Jun 13, 2018 at 11:17:18AM -0700, ftynse wrote:
> No, I wrote a completely new `isPromotableToRegistersBelow` (including the unrolling par
t that I separated in a different commit) and then called it from `isPromotableToRegisters
BelowThreads`.

Let's try and avoid reimplementing the same functionality.
I removed these changes.  If you would still like
to split up this function, then please do it 
in a separate PR.